### PR TITLE
The OverscrollIndicator should not overflow the scrollable view's edge

### LINKED ISSUE #64149: ListView highlight move above other Widgets
## Description

Hello, I have a `Widget` at the top of a `ListView`. Both are the children of a `Column`.
And when I scroll down and then I scroll up quickly the highlight of the `ListView` move above the `Widget` at the top.
You can see this with the image under.

Demo

## Steps to Reproduce

Use the following code to reproduce the issue.

``` dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Flutter demo"),
      ),
      body: Column(
        children: <Widget>[
          Container(
            height: 100,
            width: double.infinity,
            color: Colors.green,
            child: Center(
              child: Text("Some widget"),
            ),
          ),
          Expanded(
            child: ListView.builder(
              itemExtent: 50,
              itemCount: 20,
              itemBuilder: (BuildContext context, int index) => Container(
                color: Colors.red,
                child: Center(
                  child: Text("Item $index"),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
```

<details>

<summary>Devices Tested</summary>

Tested on Android 5.1.1 and Web.

</details>

<details>

<summary>Flutter Doctor</summary>

```
[✓] Flutter (Channel beta, 1.20.2, on Linux, locale en_US.UTF-8)
    • Flutter version 1.20.2 at /home/mpi3d/flutter
    • Framework revision bbfbf1770c (6 days ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /home/mpi3d/Android/Sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (2 available)
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 84.0.4147.125

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by xu-baolin**
## Description

The OverscrollIndicator can not overflow the scrollable edge.

Related commits:
https://github.com/flutter/flutter/commit/362a5573651f6fb7ab1e7e16b872de62ff1fea67
https://github.com/flutter/flutter/commit/39e00d2ea4991b44027d501e32243c27a6268f47

## Related Issues

Fixes https://github.com/flutter/flutter/issues/64149

## Tests

I added the following tests:

`CustomScrollView overscroll indicator works if there is sliver before center and modify glow position`

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by xu-baolin** [comment]
@chunhtai @Piinks Please review, thanks:)

**Comment by xu-baolin** [comment]
@HansMuller Hi, Please review this change if it is convenient for you, thank you very much :)

File: packages/flutter/test/widgets/overscroll_indicator_test.dart [hunk 1]
```diff
@@ -383,13 +430,25 @@ void main() {
...
       expect(painter, paints..save()..translate(y: 50.0)..scale()..circle());
-      // Reverse scroll direction.
+      // Reverse scroll (30 pixels), and the offset < notification.paintOffset.
       await tester.dragFrom(const Offset(200.0, 200.0), const Offset(0.0, -30.0));
       await tester.pump();
-      // The painter should follow the scroll direction.
+      // The painter should follow the scroll direction .
       expect(painter, paints..save()..translate(y: 50.0 - 30.0)..scale()..circle());
+
+      // Regressing test for https://github.com/flutter/flutter/issues/64149
```
Comment (chunhtai) [inline packages/flutter/test/widgets/overscroll_indicator_test.dart:441]: Can you separate this into a different test?

File: packages/flutter/test/widgets/overscroll_indicator_test.dart [hunk 2]
```diff
@@ -416,11 +475,23 @@ void main() {
...
-      // Reverse scroll direction.
+      // Reverse scroll (30 pixels).
       await tester.dragFrom(const Offset(200.0, 200.0), const Offset(0.0, 30.0));
       await tester.pump();
       // The painter should follow the scroll direction.
       expect(painter, paints..scale(y: -1.0)..save()..translate(y: 50.0 - 30.0)..scale()..circle());
+
+      // Regressing test for https://github.com/flutter/flutter/issues/64149
+      // Reverse scroll (30+20 pixels) and offset == notification.paintOffset.
+      await tester.dragFrom(const Offset(200.0, 200.0), const Offset(0.0, 20.0));
```
Comment (chunhtai) [inline packages/flutter/test/widgets/overscroll_indicator_test.dart:486]: same here

File: packages/flutter/test/widgets/overscroll_indicator_test.dart [hunk 3]
```diff
@@ -362,6 +362,53 @@ void main() {
...
+        ),
+      ),
+    );
+
+    expect(find.text('First sliver 1'), findsNothing);
+
+    await slowDrag(tester, const Offset(200.0, 200.0), const Offset(0.0, 300.0));
+    expect(find.text('First sliver 1'), findsOneWidget);
+    final RenderObject painter = tester.renderObject(find.byType(CustomPaint));
+    // The painter should follow the scroll direction.
```
Comment (chunhtai) [inline packages/flutter/test/widgets/overscroll_indicator_test.dart:408]: I thought we are verifying the paint offset is correctly applied, what does the scroll direction do here?

**Review by chunhtai — COMMENTED** [review]
Sorry for the late reply. change looks good to me, i left some comments on tests.

File: packages/flutter/test/widgets/overscroll_indicator_test.dart (hunk 3 above)
Comment (xu-baolin) [reply packages/flutter/test/widgets/overscroll_indicator_test.dart:408]: Sorry for the comment is not clear, I will fix it.
Line 375 set the `paintOffset` 50.0, `[OverscrollIndicatorNotification.paintOffset]` can special the indicator offset, also see https://api.flutter.dev/flutter/widgets/OverscrollIndicatorNotification/paintOffset.html
So in this test case, the indicator should paint at 50.0 from its edge.

File: packages/flutter/test/widgets/overscroll_indicator_test.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/test/widgets/overscroll_indicator_test.dart]: Done.

File: packages/flutter/test/widgets/overscroll_indicator_test.dart (hunk 2 above)
Comment (xu-baolin) [reply packages/flutter/test/widgets/overscroll_indicator_test.dart]: Done.

File: packages/flutter/test/widgets/overscroll_indicator_test.dart [hunk 4]
```diff
@@ -355,13 +355,122 @@ void main() {
 
     expect(find.text('First sliver 1'), findsNothing);
 
-    await slowDrag(tester, const Offset(200.0, 200.0), const Offset(0.0, 300.0));
+    await slowDrag(tester, const Offset(200.0, 200.0), const Offset(0.0, 5.0)); // offset will be magnified ten times
```
Comment (xu-baolin) [inline packages/flutter/test/widgets/overscroll_indicator_test.dart:358]: The `slowDrag` will ten times the offset, so just drag 50.0 pixels can trigger the overscroll indicator.

This all looks fine however this method occurs about 18 times in this test and the `slowDrag` parameter is always small value. Probably best to bow to convention and do the same here. :)
image

File: packages/flutter/test/widgets/overscroll_indicator_test.dart [hunk 5]
```diff
@@ -362,6 +362,115 @@ void main() {
...
+        ),
+      ),
+    );
+    final RenderObject painter = tester.renderObject(find.byType(CustomPaint));
+    await slowDrag(tester, const Offset(200.0, 200.0), const Offset(0.0, 5.0));
+    expect(painter, paints..save()..translate(y: 50.0)..scale()..circle());
+    // Reverse scroll (30 pixels), and the offset < notification.paintOffset.
+    await tester.dragFrom(const Offset(200.0, 200.0), const Offset(0.0, -30.0));
+    await tester.pump();
+    // The painter should follow the scroll direction.
```
Comment (chunhtai) [inline packages/flutter/test/widgets/overscroll_indicator_test.dart:436]: comment is a bit confusing

File: packages/flutter/test/widgets/overscroll_indicator_test.dart [hunk 6]
```diff
@@ -362,6 +362,115 @@ void main() {
...
+    await tester.pumpAndSettle(); // Finish the leading indicator.
+
+    // trigger the trailing indicator
+    await slowDrag(tester, const Offset(200.0, 200.0), const Offset(0.0, -200.0));
+    expect(painter, paints..scale(y: -1.0)..save()..translate(y: 50.0)..scale()..circle());
+
+    // Reverse scroll (30 pixels), and the offset < notification.paintOffset.
+    await tester.dragFrom(const Offset(200.0, 200.0), const Offset(0.0, 30.0));
+    await tester.pump();
+    // The painter should follow the scroll direction.
```
Comment (chunhtai) [inline packages/flutter/test/widgets/overscroll_indicator_test.dart:459]: same here

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/test/widgets/overscroll_indicator_test.dart (hunk 5 above)
Comment (xu-baolin) [reply packages/flutter/test/widgets/overscroll_indicator_test.dart]: Fixed.

File: packages/flutter/test/widgets/overscroll_indicator_test.dart (hunk 6 above)
Comment (xu-baolin) [reply packages/flutter/test/widgets/overscroll_indicator_test.dart]: Fixed.

**Pull request merged**